src/coroutine.rs

```diff
@@ -228,15 +239,33 @@ impl Coroutine {
     //
     // Tracking issue: https://github.com/zonyitoo/coio-rs/issues/51
     #[doc(hidden)]
-    #[inline(never)]
+    #[inline(always)]
     pub fn yield_with(&mut self, state: State, data: usize) -> usize {
         let context = self.take_context();
 
         trace!("{:?}: yielding to {:?}", self, &context);
 
         self.state = state;
 
+        // yield ...
         let Transfer { context, data } = context.resume(data);
+        // ...continue
+
+        // Check if it is time to unwind
+        if self.unwind_flag.load(::std::sync::atomic::Ordering::Acquire) {
```

Comment:
If you read my other comment about passing `usize::max_value()`:
Here you could now check for `data == usize::max_value()` (generated by the `drop()` method) and in that case unwind the stack.
The data value is currently only used to pass a function pointer to the `Processor` for the `Processor::park_with()` method (which might get replaced with `resume_ontop()` in the future anyways).
It will thus never yield a invalid address like `0` or `usize::max_value()` (those are errors values in POSIX anyways).

BTW: You don't need to synchronize internal state using atomics. Coroutines are inherently not thread safe and the current thread owns them to it's full extend. Whenever Coroutines are transferred from one thread to another the memory access will be synchronized using atomics/barriers anyways already.